src/state/selectors/barSelectors.ts

```
@@ -0,0 +1,495 @@
+
+const pickYAxisId = (_state: RechartsRootState, _xAxisId: AxisId, yAxisId: AxisId): AxisId => yAxisId;
+
+const pickBarSettings = (
+  _state: RechartsRootState,
+  _xAxisId: AxisId,
+  _yAxisId: AxisId,
+  _isPanorama: boolean,
+  barSettings: BarSettings,
+): BarSettings => barSettings;
+
+const pickCells = (
+  _state: RechartsRootState,
+  _xAxisId: AxisId,
+  _yAxisId: AxisId,
+  _isPanorama: boolean,
+  _barSettings: BarSettings,
+  cells: ReadonlyArray<ReactElement> | undefined,
+): ReadonlyArray<ReactElement> | undefined => cells;
+
+const getBarSize = (
+  globalSize: number | undefined,
+  totalSize: number | undefined,
+  selfSize: number | string | undefined,
+): number | undefined => {
+  const barSize: string | number | undefined = isNil(selfSize) ? globalSize : selfSize;
+
+  return isNil(barSize) ? undefined : getPercentValue(barSize, totalSize, 0);
+};
+
+export const selectAllVisibleBars: (
+  state: RechartsRootState,
+  xAxisId: AxisId,
+  yAxisId: AxisId,
+) => ReadonlyArray<CartesianGraphicalItemSettings> = createSelector(
+  [selectChartLayout, selectUnfilteredCartesianItems, pickXAxisId, pickYAxisId],
+  (layout: LayoutType, allItems, xAxisId, yAxisId) =>
+    allItems
+      .filter(i => {
+        if (layout === 'horizontal') {
+          return i.xAxisId === xAxisId;
+        }
+        return i.yAxisId === yAxisId;
+      })
+      .filter(i => i.hide === false)
+      .filter(i => i.type === 'bar'),
+);
+
+type BarCategory = {
+  stackId: StackId | undefined;
+  /**
+   * List of dataKeys of items stacked at this position.
+   * All of these Bars are either sharing the same stackId,
+   * or this is an array with one Bar because it has no stackId defined.
+   *
+   * This structure limits us to having one dataKey only once per stack which I think is reasonable.
+   * People who want to have the same data twice can duplicate their data to have two distinct dataKeys.
```

What happens if someone does try to do this? Silent failure? explosion?